framework/framework.go

```diff
@@ -142,3 +142,60 @@ func (f *Framework) ProcessDelete(obj interface{}, resources []Resource) error {
 
 	return nil
 }
+
+// ProcessUpdate is a drop-in for an informer's UpdateFunc. It receives the
+// custom object observed during TPR watches and anything that implements
+// Resource. ProcessUpdate takes care about all necessary reconciliation logic
+// for update events. For complex resources this means state has to be created,
+// deleted and updated eventually, in this order.
+//
+//     func updateFunc(oldObj, newObj interface{}) {
+//         err := f.ProcessUpdate(oldObj, newObj, resources)
+//         if err != nil {
+//             // error handling here
+//         }
+//     }
+//
+//     newResourceEventHandler := &cache.ResourceEventHandlerFuncs{
+//         UpdateFunc:    updateFunc,
+//     }
+//
+func (f *Framework) ProcessUpdate(oldObj, newObj interface{}, resources []Resource) error {
+	if len(resources) == 0 {
+		return microerror.Maskf(executionFailedError, "resources must not be empty")
+	}
+
+	for _, r := range resources {
+		currentState, err := r.GetCurrentState(oldObj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		desiredState, err := r.GetDesiredState(newObj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		createState, deleteState, updateState, err := r.GetUpdateState(newObj, currentState, desiredState)
```

Comment:
The update state is concerned about creating, deleting and updating. Lets consider the example of our guest clusters. We have a TPO/CRO for a guest cluster. The TPO/CRO might be updated by adding, removing or modifying the settings of a worker within the guest cluster. Therefore different operations have to be applied to different state objects. When adding a worker to the cluster TPO/CRO the resource reconciliation has to add a worker resource on the update event. When removing a worker from the cluster TPO/CRO the resource reconciliation has to remove a worker resource on the update event. When modifying a worker in the cluster TPO/CRO the resource reconciliation has to modify a worker resource on the update event. It is not good enough to reuse functionality as implemented by e.g. `GetDeleteState` because this is concerned about deleting all data. Deletion on update events is concerned with partial removal of parts of the resource being reconciled.